src/utils/AssayGeneratorMetabo.cpp

```diff
@@ -132,157 +127,110 @@ class TOPPAssayGeneratorMetabo :
+  struct compare
   {
-    map<const BaseFeature*, vector<size_t>> feature_ms2_spectra_map;
-
-    // map precursors to closest feature and retrieve annotated metadata (if possible)
-    for (size_t index = 0; index != spectra.size(); ++index)
+  	string key;
+  	compare(string const &i): key(i){}
+  
+  	bool operator()(string const &i)
+  	{
+  		return (i == key);
+  	}
+  };
+
+  static bool compoundNameLess_(const OpenMS::MetaboTargetedAssay& i, const OpenMS::MetaboTargetedAssay& j)
+  {
+    if (i.compound_name < j.compound_name)
     {
-      if (spectra[index].getMSLevel() != 2) { continue; }
-
-      // get precursor meta data (m/z, rt)
-      const vector<Precursor> & pcs = spectra[index].getPrecursors();
-
-      if (!pcs.empty())
+      return (i.compound_name < j.compound_name);
+    }
+    if (i.compound_name == j.compound_name)
+    {
+      if (i.compound_adduct == j.compound_adduct)
       {
-        const double mz = pcs[0].getMZ();
-        const double rt = spectra[index].getRT();
-
-        // query features in tolerance window
-        vector<Size> matches;
-
-        // get mz tolerance window
-        std::pair<double,double> mz_tolerance_window = Math::getTolWindow(mz, precursor_mz_tolerance, ppm);
-        fp_map_kd.queryRegion(rt - precursor_rt_tolerance, rt + precursor_rt_tolerance, mz_tolerance_window.first, mz_tolerance_window.second, matches, true);
-
-        // no precursor matches the feature information found
-        if (matches.empty()) { continue; }
-
-        // in the case of multiple features in tolerance window, select the one closest in m/z to the precursor
-        Size min_distance_feature_index(0);
-        double min_distance(1e11);
-        for (auto const & k_idx : matches)
-        {
-          const double f_mz = fp_map_kd.mz(k_idx);
-          const double distance = fabs(f_mz - mz);
-          if (distance < min_distance)
-          {
-            min_distance = distance;
-            min_distance_feature_index = k_idx;
-          }
-        }
-        const BaseFeature* min_distance_feature = fp_map_kd.feature(min_distance_feature_index);
-
-        feature_ms2_spectra_map[min_distance_feature].push_back(index);
+        return (i.precursor_int > j.precursor_int);
+      }
+      else
+      {
+        return (i.compound_adduct < j.compound_adduct);
       }
     }
-    return feature_ms2_spectra_map;
+    else
+    {
+      return false;
+    }
   }
-
-  // precursor correction (highest intensity)
-  Int getHighestIntensityPeakInMZRange(double test_mz, const MSSpectrum& spectrum1, double tolerance, bool ppm)
+ 
+  // use std::unique with this function to retain the unique element with the highest precursor intensity
+  static bool compoundUnique_(const OpenMS::MetaboTargetedAssay& i, const OpenMS::MetaboTargetedAssay& j)
```

Comment:
https://en.cppreference.com/w/cpp/algorithm/unique: Elements are compared using the given binary predicate p. The behavior is undefined if it is not an equivalence relation.

Reply:
Yes, I will change that.